trezor: support new passphrase entry mechanism, requires #6446

ph4r05 opened the pull request:
- device/host decision is now made on the host (due to the change in Trezor firmware)
- passphrase can be prompted also when wallet is initialized (thus I had to add passphrase support to the wallet listener)
- requires https://github.com/monero-project/monero/pull/6446 to work

GUI now asks where does user want to enter the passphrase hardware wallet or host.
<img width="979" alt="Screenshot 2020-04-12 at 19 42 33" src="https://user-images.githubusercontent.com/1052761/79076250-f4433400-7cf8-11ea-9880-775c641c7856.png">

If host is selected, passphrase prompt is displayed
<img width="978" alt="Screenshot 2020-04-12 at 19 39 27" src="https://user-images.githubusercontent.com/1052761/79076248-f1e0da00-7cf8-11ea-9108-676147fa643d.png">


If device is selected the app behaves in the same way as previously. 
<img width="977" alt="Screenshot 2020-04-11 at 12 32 09" src="https://user-images.githubusercontent.com/1052761/79072604-3f058180-7ce2-11ea-97c6-e863695e870e.png">

Unrelated to this, but https://github.com/monero-project/monero/pull/6446 fixes also missing fee information:
<img width="523" alt="Screenshot 2020-04-11 at 13 49 46" src="https://user-images.githubusercontent.com/1052761/79072608-42990880-7ce2-11ea-9f2f-5fe92440c35c.png">

selsta commented:
Can you blur the background? https://github.com/monero-project/monero-gui/blob/86252506f01a8069ffbf01bed2ed8f44123c5233/main.qml#L1694

ph4r05 commented:
> Can you blur the background when passphraseDialog is visible?
> 
> https://github.com/monero-project/monero-gui/blob/86252506f01a8069ffbf01bed2ed8f44123c5233/main.qml#L1694

Thanks for the suggestion. I changed it, but now I realized I have to make small adjustments to the passphrase logic so I switched PR to draft.

ph4r05 commented:
Done

ph4r05 commented:
Ah. I've noticed QScopeGuard is available only since Qt 5.12. I will address this.

ph4r05 commented:
@selsta CI builds are failing as PR monero dependency is not resolved. 
Any idea why this is not working in CI?

https://github.com/monero-project/monero-gui/blob/cd9ca2daecdf3d81c4c7d65de79e97f2387d6942/get_libwallet_api.sh#L38

selsta commented:
I don’t know but once the PR is merged you can force push to rerun the CI.

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2014-2019, The Monero Project
```
2020 :)

ph4r05 commented:
> * Can you remove the black rectangle box from here? Our password dialogs usually only have blur with no background.
> Can you remove the black rectangle box from here? Our password dialogs usually only have blur with no background.

@selsta do you mean background for password box or background around the whole dialog?

selsta commented:
@ph4r05 The black rectangle background for the password box. The StandardDialog is okay.

selsta commented:
<img width="1026" alt="Screenshot 2020-05-05 at 21 48 22" src="https://user-images.githubusercontent.com/7697454/81109280-258ada80-8f1a-11ea-9308-c74d2a3bc4a7.png">
Our password entry looks like this

ph4r05 commented:
@selsta like this?
<img width="753" alt="Screenshot 2020-05-06 at 09 27 23" src="https://user-images.githubusercontent.com/1052761/81148199-20f70e00-8f7c-11ea-9152-00b750e4c604.png">
<img width="767" alt="Screenshot 2020-05-06 at 09 27 29" src="https://user-images.githubusercontent.com/1052761/81148205-22c0d180-8f7c-11ea-8ed7-3ac6548679b6.png">

ph4r05 commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,158 @@
+            }
+        }
+    }
+}
```
done

rating89us commented:
Your password textboxes are still black. We use it transparent with blur, see: 
![image](https://user-images.githubusercontent.com/45968869/81149743-15f1ad00-8f7f-11ea-9497-d4507da91edf.png)

selsta commented:
@ph4r05 Can you try to rebase with master? This should make them transparent.

ph4r05 commented:
> Your password textboxes are still black. We use it transparent with blur, see:

Isn't it QT thing? I see it this way:
<img width="663" alt="Screenshot 2020-05-06 at 09 55 00" src="https://user-images.githubusercontent.com/1052761/81150123-b21bb400-8f7f-11ea-9be2-8e2bb21ed668.png">

ph4r05 commented:
This PR should not affect theme of other components (like PasswordDialog)

selsta commented:
> This PR should not affect the theme of other components (like PasswordDialog)

Correct. Once rebased it looks like the other screenshots:

<img width="1026" alt="Screenshot 2020-05-06 at 10 22 32" src="https://user-images.githubusercontent.com/7697454/81153743-83074180-8f83-11ea-8987-d1e1879589e7.png">

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+        cancelText: qsTr("Host")
+        closeVisible: true
+        icon: StandardIcon.Question
+        property var isClosedWithoutAction: false
```
```suggestion
```

ph4r05 replied:
done

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+        okText: qsTr("Hardware wallet")
+        cancelText: qsTr("Host")
+        closeVisible: true
+        icon: StandardIcon.Question
```
```suggestion
```

ph4r05 replied:
done

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+        okVisible: true
+        okText: qsTr("Hardware wallet")
+        cancelText: qsTr("Host")
+        closeVisible: true
```
```suggestion
```

ph4r05 replied:
done

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+        title: qsTr("Hardware wallet passphrase") + translationManager.emptyString
+        text: qsTr("Please select where you want to enter passphrase.\nIt is recommended to enter passphrase on the hardware wallet for better security.") + translationManager.emptyString
+        cancelVisible: true
+        okVisible: true
```
```suggestion
```

ph4r05 replied:
done

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+        id: entryChooserDialog
+        title: qsTr("Hardware wallet passphrase") + translationManager.emptyString
+        text: qsTr("Please select where you want to enter passphrase.\nIt is recommended to enter passphrase on the hardware wallet for better security.") + translationManager.emptyString
+        cancelVisible: true
```
```suggestion
```

ph4r05 replied:
done

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+            onAcceptedCallback(passphraseDialog.password);
+        }
+        onRejectedPassphrase: {
+        if (onRejectedCallback)
```
indentation looks wrong, also above

ph4r05 replied:
done

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+
+    PasswordDialog {
+        id: passphraseDialog
+        z: root.z + 1
```
```suggestion
```

ph4r05 replied:
done

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import QtQuick 2.9
+import QtQuick.Controls 2.0
```
Remove imports: `QtQuick.Controls 2.0`, `QtQuick.Dialogs 1.2`, `QtQuick.Layouts 1.1`, `QtQuick.Controls.Styles 1.4`, `QtQuick.Window 2.0` `FontAwesome 1.0`, `"effects/" as MoneroEffects`, `"../js/Utils.js" as Utils`

ph4r05 replied:
done

ph4r05 commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+    anchors.fill: parent
+    z: parent.z + 1
+
+    property var canEnterOnDevice
```
done

ph4r05 commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    StandardDialog {
+        z: root.z + 1
```
done

ph4r05 commented:
@selsta thanks for a great feedback!

btw I've just figured out that StandardDialog resets the ok/cancel button labels on close. Any ideas where to set 

```
okText: qsTr("Hardware wallet")
cancelText: qsTr("Host")
```

variables so it is set correctly? Override open()?

selsta commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,106 @@
+        root.visible = true;
+
+        if (canEnterOnDevice) {
+            entryChooserDialog.open()
```
You can set the `okText` and `cancelText` here and remove them from the component itself.

ph4r05 replied:
done

selsta commented:
@xiphon has to review the C++ part

selsta commented on main.qml:
```diff
@@ -1537,14 +1550,13 @@ ApplicationWindow {
+    DevicePassphraseDialog {
+        id: devicePassphraseDialog
+        visible: false
+        z: parent.z + 1
```
```suggestion
```

selsta commented on main.qml:
```diff
@@ -1537,14 +1550,13 @@ ApplicationWindow {
+        id: devicePassphraseDialog
+        visible: false
+        z: parent.z + 1
+        anchors.fill: parent
```
```suggestion
```

xiphon commented on src/libwalletqt/Wallet.h:
```diff
@@ -348,6 +349,10 @@ class Wallet : public QObject
 
+    // Passphrase entry for hardware wallets
+    Q_INVOKABLE void onPassphraseEntered(const QString &passphrase, bool enter_on_device, bool entry_abort=false);
+    Q_INVOKABLE virtual void onWalletPassphraseNeeded(bool on_device) override;
```
Do you plan to invoke this from QML?

ph4r05 replied:
Currently, `onPassphraseEntered` is called by the QML, `onWalletPassphraseNeeded` is not (is called by passphrase helper from the wallet listener, causes signal emission to the QML, displays prompt). Should I remove `Q_INVOKABLE` from `onWalletPassphraseNeeded` then?

xiphon replied:
Yep, would prefer removing `Q_INVOKABLE ` from `onWalletPassphraseNeeded` if we don't call it from QML.

ph4r05 replied:
Done, pls check

xiphon replied:
Looks okay to me.
Only the PR author can `Resolve` it.

xiphon commented on src/libwalletqt/PassphraseHelper.cpp:
```diff
@@ -0,0 +1,78 @@
+    qDebug() << __FUNCTION__;
+    onWalletPassphraseNeeded(on_device);
+
+    if (m_passphrase_abort)
+    {
+        throw std::runtime_error("Passphrase entry abort");
+    }
+
+    on_device = m_passphrase_on_device;
+    if (!on_device) {
+        auto tmpPass = m_passphrase.toStdString();
+        m_passphrase = QString();
+        return Monero::optional<std::string>(tmpPass);
+    } else {
+        return Monero::optional<std::string>();
+    }
```
Looks like this should be guarded by the mutex as well, no?

ph4r05 replied:
After the `onWalletPassphraseNeeded` returns the passphrase entry is finished and we just return the value and set `on_device`, I think the mutex is not needed. 
I use mutex in the PassphraseHelper just to guard state when using `QWaitCondition` to make sure it works properly.

xiphon replied:
`onDevicePassphraseRequest` and `onPassphraseEntered` are concurrent. Have a good chance to corrupt the state then.

ph4r05 replied:
I think I still don't see the scenario that could cause the corruption. 
The call flow goes like 

- `thread-wallet-api`: `onDevicePassphraseRequest` -> `onWalletPassphraseNeeded` mutex & cond wait for UI, wait
- `thread-gui`: UI calls `onPassphraseEntered`, fills in data, wakes condition variable which causes unblocking of the `thread-wallet-api`
- `thread-wallet-api`: continues after line `onWalletPassphraseNeeded`, returns eventually. 

For a race to happen GUI would have to submit another `onPassphraseEntered` somehow out of order. 

To be extra safe we could add 

```cpp
    m_mutex_pass.lock();
    auto cleanup = qScopeGuard([&] { m_mutex_pass.unlock; });
    if (m_passphrase_abort)
    {
        throw std::runtime_error("Passphrase entry abort");
    }

    on_device = m_passphrase_on_device;
    if (!on_device) {
        auto tmpPass = m_passphrase.toStdString();
        m_passphrase = QString();
        return Monero::optional<std::string>(tmpPass);
    } else {
        return Monero::optional<std::string>();
    }
```

What do you think?

xiphon replied:
Looks better, I would use `QMutexLocker` instead of `qScopeGuard` though.

What do you think about dropping `onWalletPassphraseNeeded` (moving it to `onDevicePassphraseRequest`)?

ph4r05 replied:
Done, pls check

xiphon replied:
Seems fine to me now

xiphon commented on src/libwalletqt/PassphraseHelper.h:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    ~QScopeGuard()
+    {
+        if (m_invoke)
+            m_func();
+    }
+
+    void dismiss() noexcept
+    {
+        m_invoke = false;
+    }
+
+private:
+    explicit QScopeGuard(F &&f) noexcept
+        : m_func(std::move(f))
+    {
+    }
+
+    Q_DISABLE_COPY(QScopeGuard)
+
+    F m_func;
+    bool m_invoke = true;
+    friend QScopeGuard qScopeGuard<F>(F);
+};
+
+
+template <typename F>
+#if QT_HAS_CPP_ATTRIBUTE(nodiscard)
+Q_REQUIRED_RESULT
+#endif
+QScopeGuard<F> qScopeGuard(F f)
+{
+    return QScopeGuard<F>(std::move(f));
+}
+
+QT_END_NAMESPACE
+
+#endif // QSCOPEGUARD_H
+#endif
```
Please move `QScopeGuard` stuff to a separate header in `src/qt`

ph4r05 replied:
Done, pls check and resolve when ok.

xiphon replied:
Seems fine to me now. Feel free to `Resolve` it.

xiphon commented on src/libwalletqt/Wallet.cpp:
```diff
@@ -1021,6 +1037,20 @@ void Wallet::keyReuseMitigation2(bool mitigation)
+
+void Wallet::onPassphraseEntered(const QString &passphrase, bool enter_on_device, bool entry_abort)
+{
+    PassphraseDelegate * pdelegate = dynamic_cast<PassphraseDelegate*>(m_walletListener);
```
`dynamic_cast` is not needed here.

ph4r05 replied:
without it I cannot use the `PassphraseDelegate`: `Monero::WalletListener *m_walletListener;` 

The `m_walletListener` is instantiated with `class WalletListenerImpl : public  Monero::WalletListener, public PassphraseDelegate` but without dynamic cast I think I cannot reach the `PassphraseDelegate`

xiphon replied:
Do you see reason to not make it `WalletListenerImpl *m_walletListener;`?

ph4r05 replied:
We would have to move `WalletListenerImpl` out of the `Wallet.cpp`, probably to a separate `.h, .cpp` files.

xiphon replied:
That's okay

ph4r05 replied:
Done, pls check and resolve when ok.

xiphon replied:
Looks okay to me

xiphon commented on src/libwalletqt/PassphraseHelper.h:
```diff
@@ -0,0 +1,149 @@
+    virtual void onPassphraseEntered(const QString &passphrase, bool enter_on_device, bool entry_abort) = 0;
+};
+
+class PassphraseHelper {
```
Shouldn't it inherit `PassphraseHandler` and  `PassphraseDelegate`?

ph4r05 replied:
PassphraseHelper is only a simple class wrapping mutex/condition variable logic to wait until user enters the passphrase (as passphrase entry is blocking in the Monero wallet API).

the `PassphraseHandler` is then implemented by `{Wallet, WalletManager}` as they are responsible for showing passphrase prompt to the user, while

the `PassphraseDelegate` is implemented by Monero wallet listener as it is responsible for receiving passphrase from the user and passing it back to the monero api.

ph4r05 replied:
Maybe we can rename the interfaces `PassphraseHandler -> PassprasePrompter` and `PassphraseDelegate -> PassphraseReceiver`, is it more understandable?

xiphon replied:
> `PassphraseHandler -> PassprasePrompter` and `PassphraseDelegate -> PassphraseReceiver`, is it more understandable?

Sounds good

ph4r05 replied:
Done, pls check and resolve when ok.

xiphon replied:
Seems fine to me now

xiphon commented on src/libwalletqt/PassphraseHelper.cpp:
```diff
@@ -0,0 +1,78 @@
+        m_handler->onWalletPassphraseNeeded(on_device);
+    }
+
+    qDebug() << __FUNCTION__ << " entered";
```
Can't find any benefit of having this debug message.

xiphon commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,106 @@
+
+Item {
+    id: root
+    anchors.fill: parent
+    z: parent.z + 1
```
```suggestion
```

ph4r05 replied:
done

xiphon commented on src/libwalletqt/PassphraseHelper.cpp:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    m_cond_pass.wait(&m_mutex_pass);
+    locker.relock();
```
Is not needed

ph4r05 replied:
done

xiphon commented on src/qt/ScopeGuard.h:
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2014-2020, The Monero Project
```
We are not the authors of original code from Qt

ph4r05 replied:
done

xiphon replied:
Hm. I still think we shouldn't put Monero copyright here.

xiphon commented on src/libwalletqt/WalletManager.h:
```diff
@@ -185,14 +186,14 @@ class WalletManager : public QObject
-    Q_INVOKABLE void onWalletPassphraseNeeded(Monero::Wallet * wallet);
-    Q_INVOKABLE void onPassphraseEntered(const QString &passphrase, bool entry_abort=false);
+    Q_INVOKABLE void onPassphraseEntered(const QString &passphrase, bool enter_on_device, bool entry_abort=false);
+    Q_INVOKABLE virtual void onWalletPassphraseNeeded(bool on_device) override;
```
This is not getting invoked from QML too

ph4r05 replied:
done

selsta commented on src/qt/ScopeGuard.h:
```diff
@@ -0,0 +1,77 @@
+#ifndef MONERO_GUI_SCOPEGUARD_H
```
Now that xiphon mentioned it, might be a good idea to add following to the top:

```
Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com, author Sérgio Martins <sergio.martins@kdab.com>
Contact: https://www.qt.io/licensing/
```

xiphon replied:
I would say you should copy the whole license header from the original Qt code.

ph4r05 replied:
Heh this becomes a problem now. The license is https://www.gnu.org/licenses/lgpl-3.0.html
Is it compatible with Monero-GUI?

selsta replied:
```
/****************************************************************************
**
** Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com, author Sérgio Martins <sergio.martins@kdab.com>
** Contact: https://www.qt.io/licensing/
**
** This file is part of the QtCore module of the Qt Toolkit.
**
** $QT_BEGIN_LICENSE:LGPL$
** Commercial License Usage
** Licensees holding valid commercial Qt licenses may use this file in
** accordance with the commercial license agreement provided with the
** Software or, alternatively, in accordance with the terms contained in
** a written agreement between you and The Qt Company. For licensing terms
** and conditions see https://www.qt.io/terms-conditions. For further
** information use the contact form at https://www.qt.io/contact-us.
**
** GNU Lesser General Public License Usage
** Alternatively, this file may be used under the terms of the GNU Lesser
** General Public License version 3 as published by the Free Software
** Foundation and appearing in the file LICENSE.LGPL3 included in the
** packaging of this file. Please review the following information to
** ensure the GNU Lesser General Public License version 3 requirements
** will be met: https://www.gnu.org/licenses/lgpl-3.0.html.
**
** GNU General Public License Usage
** Alternatively, this file may be used under the terms of the GNU
** General Public License version 2.0 or (at your option) the GNU General
** Public license version 3 or any later version approved by the KDE Free
** Qt Foundation. The licenses are as published by the Free Software
** Foundation and appearing in the file LICENSE.GPL2 and LICENSE.GPL3
** included in the packaging of this file. Please review the following
** information to ensure the GNU General Public License requirements will
** be met: https://www.gnu.org/licenses/gpl-2.0.html and
** https://www.gnu.org/licenses/gpl-3.0.html.
**
** $QT_END_LICENSE$
**
****************************************************************************/
```

appears to be the full license

https://github.com/RSATom/Qt/blob/master/qtbase/src/corelib/tools/qscopeguard.h

selsta replied:
looks ok?

https://opensource.stackexchange.com/questions/6405/lgpl-licensed-code-reuse-in-mit-licensed-project

ph4r05 replied:
This seems problematic at this point. We like to have something like scope guard to have RAII resource management, but without upgrading the Qt I think we cannot use this source without forcing the whole Monero-GUI to be LGPLv3, am I right?

Do we need to reimplement scope guard by ourselves? And if it is the case, it probably cannot be called `qScopeGuard` anymore as it won't have the same properties. 
Then we need to have something ours, which fallbacks to `qScopeGuard` if it is present. Am I right?

xiphon replied:
> Heh this becomes a problem now. The license is https://www.gnu.org/licenses/lgpl-3.0.html
> Is it compatible with Monero-GUI?

At the first glance, we ain't doing anything malicious, nor we violate the license.

We don't claim the copyright on the code and we copy the original license and copyright.

selsta replied:
There is also this: https://spdx.org/licenses/Qt-LGPL-exception-1.1.html

ph4r05 replied:
I am removing this scope guard and using https://github.com/ricab/scope_guard

ph4r05 replied:
Any ideas where to put the scope guard related files? it does not seem right to put it in `src/qt`, should I create `src/misc` ?

selsta replied:
FWIW we already have a Qt file in the source code tree (MoneroSettings) so if we violate the license we did for a while.

xiphon replied:
> I am removing this scope guard and using https://github.com/ricab/scope_guard

Great, that is a good one.

>Any ideas where to put the scope guard related files? it does not seem right to put it in `src/qt`, should I create `src/misc` ?

That's correct. But we already have some unrelated things in `src/qt`, please place it there for now.

ph4r05 replied:
Hmm I still think that having LGPL sources in the app is problematic. This LGPL has to be stated somewhere in the app (about), the app is distributed also in the binary forms. I am not sure whether this is safe. LGPL states 

> A program that contains no derivative of any portion of the Library, but is designed to work with the Library by being compiled or linked with it, is called a "work that uses the Library". Such a work, in isolation, is not a derivative work of the Library, and therefore falls outside the scope of this License.

I think that means that whole Monero-GUI would have to be LGPL when we use the source file directly. https://www.quora.com/If-my-code-uses-some-code-licensed-under-LGPL-can-I-release-the-whole-software-under-MIT

ph4r05 replied:
So what do you think guys? I would strongly prefer to use Unlicensed https://github.com/ricab/scope_guard to avoid such license pains. Do you agree?

If yes, should I create `src/misc` or `src/utils` where to put these new files or should I use some existing dir?

xiphon replied:
> I would strongly prefer to use Unlicensed https://github.com/ricab/scope_guard to avoid such license pains. Do you agree?

Yes.

Already replied to this, see https://github.com/monero-project/monero-gui/pull/2829#discussion_r422211422

ph4r05 replied:
> Already replied to this, see [#2829 (comment)](https://github.com/monero-project/monero-gui/pull/2829#discussion_r422211422)

Ah I see, GitHub didn't refresh the conversation.

ph4r05 replied:
ok done, pls check

xiphon commented on src/libwalletqt/WalletManager.cpp:
```diff
@@ -539,22 +534,15 @@ WalletManager::~WalletManager()
-
-    m_cond_pass.wakeAll();
-    m_mutex_pass.unlock();
+    if (m_passphraseReceiver != nullptr)
+    {
+        m_passphraseReceiver->onPassphraseEntered(passphrase, enter_on_device, entry_abort);
```
Should protect `m_passphraseReceiver` with a mutex.

ph4r05 replied:
That it does not become null after the check?
It is set only in `WalletManager::openWallet` and `WalletManager::createWalletFromDevice`, both are protected by locking `m_mutex`. 

The scenario this could happen is: GUI somehow triggers (without previous request) `onPassphraseEntered`, meanwhile one of the two functions ends, scope guard sets `m_passphraseReceiver` to null in race. 

Should we then to protect from this add a new mutex, that is called from the scope guard, before setting `m_passphraseReceiver` to null?

xiphon replied:
> Should we then to protect from this add a new mutex, that is called from the scope guard, before setting `m_passphraseReceiver` to null?

Yes, please do

ph4r05 replied:
ok done, pls check

xiphon replied:
Is okay now

xiphon commented on components/DevicePassphraseDialog.qml:
```diff
@@ -0,0 +1,104 @@
+
+        if (canEnterOnDevice) {
+            entryChooserDialog.okText = qsTr("Hardware wallet")
+            entryChooserDialog.cancelText = qsTr("Host")
```
Could we replace `Host` with something GUI users more familiar with?

ph4r05 replied:
sure, I will set there any string you suggest

xiphon replied:
Let me summon @erciccione

erciccione replied:
I think `Computer` would be better. Sounds dumb, but it's very easy to understand for both users and translators.

xiphon commented on src/qt/ScopeGuard.h:
```diff
@@ -0,0 +1,206 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
Hm. I still think we shouldn't put Monero copyright here.

ph4r05 replied:
Its matter of preference I think, It is licensed as "Unlicense" which means we can freely relicense it to a common Monero license. We also mention the author. But I basically don't care, do you want to put there just author name as it was there previously? Or the unlicense text?

xiphon replied:
Would prefer having author name, github link and Unlicense text here

ph4r05 replied:
ok done

xiphon commented on src/libwalletqt/WalletManager.cpp:
```diff
@@ -186,6 +182,11 @@ Wallet *WalletManager::createWalletFromDevice(const QString &path, const QString
 {
     QMutexLocker locker(&m_mutex);
     WalletPassphraseListenerImpl tmpListener(this);
+    m_passphraseReceiver = &tmpListener;
```
Please guard the assignment the `m_mutex_passphraseReceiver` too

ph4r05 replied:
ah, true, done

xiphon replied:
Is okay now

xiphon commented on src/libwalletqt/WalletManager.cpp:
```diff
@@ -532,29 +537,26 @@ WalletManager::WalletManager(QObject *parent)
     , m_scheduler(this)
 {
     m_pimpl =  Monero::WalletManagerFactory::getWalletManager();
+    m_mutex_passphraseReceiver.lock();
```
The mutex is not needed here. Initializing `m_passphraseReceiver(nullptr)` will work just fine

ph4r05 commented:
Thanks @xiphon !

xiphon commented:
> Thanks @xiphon !

No problem. Thank you.
